### ISSUE TITLE
Support for encoding/decoding num_bigint type 

### ISSUE THREAD

**Issue by LouayKamel** [rusterlium/rustler#221]
Hey Everyone, Thank you all for this amazing project.

I'm working on a small algorithm that requires arithmetic operations on Arbitrary-precision numbers.
What is the ideal way to implement encode&decode methods on num_bigint::{BigInt, ..} types?

**Comment by evnu** [comment]
**TL;DR** This should be possible using binaries.

While Erlang supports [bigint](http://erlang.org/doc/efficiency_guide/advanced.html#memory), the NIF interface does not seem to support that (see [enif_get_int](http://erlang.org/doc/man/erl_nif.html#enif_get_int) and [enif_get_int64](http://erlang.org/doc/man/erl_nif.html#enif_get_int64)). So, there does not seem to be a readily available type to convert between Erlang's bigints and num_bigint. One way to approach this would be to make use of [to_signed_bytes_be()](https://docs.rs/num-bigint/0.2.2/num_bigint/struct.BigInt.html#method.to_signed_bytes_be) (or `_le`) in num_bigint. The example for `to_signed_bytes_be()` from the documentation:

```rust
use num_bigint::ToBigInt;

let i = -1125.to_bigint().unwrap();
assert_eq!(i.to_signed_bytes_be(), vec![251, 155]);
```

To decode `vec![251, 155]` in Elixir:

```elixir
vec = [251, 155]
size = length(vec) * 8
<<decoded :: signed-big-integer-size(size)>> = :erlang.list_to_binary(a)
decoded # => -1125
```

Now, I would expect this to be slow(ish), as 

* `Vec` decodes into a list
* this list needs to be copied
* its length needs to be computed
* the list must be turned into a binary
* the final binary needs to be decoded

To speed this up, the vector should be converted into a binary within the NIF. Then, decoding is simple:

```elixir
bin = <<251, 155>>
size = bit_size(bin)
<<decoded :: signed-big-integer-size(size) >> = bin
decoded # -1125
```

I think the encoder for `str` already implements most of what is needed to convert a `Vec<u8>` into a binary. Without having tested it, I believe the following could work (I replaced `str` with a newtype):

```rust
// Introduce a newtype for serialized bigint
struct BigInt(Vec<u8>);

impl Encoder for BigInt {
    fn encode<'b>(&self, env: Env<'b>) -> Term<'b> {
        let str_len = self.0.len();
        let mut bin = match OwnedBinary::new(str_len) {
            Some(bin) => bin,
            None => panic!("binary term allocation fail"),
        };
        bin.as_mut_slice()
            .write(self.as_bytes()) // or maybe write_all()
            .expect("memory copy of string failed");
        bin.release(env).to_term(env)
    }
}
```

**Comment by LouayKamel** [comment]
Thank you, this was really helpful +1
